rb_link/src/server/getter.rs

```diff
@@ -0,0 +1,190 @@
+            if let Some(b2r_message) = queue.front() {
+                if b2r_message.cycles == min_cycles {
+                    messages.push(queue.pop_front().expect("front error"));
+                }
+            }
+        }
+        messages
+    }
+}
+
+/// The pipeline state
+/// cycle: the cycle of the state
+/// full_fifos: the ids of the full fifos
+/// empty_fifos: the ids of the empty fifos
+/// fire_fules: the fired rules
+pub struct PipeLineState {
+    pub cycle: u32,
+    pub full_fifos: Vec<u32>,
+    pub empty_fifos: Vec<u32>,
+    pub fire_rules: Vec<u32>,
+}
+
+/// A getter that retrieves messages from specific probes sequentially according to cycles
+/// and aggregates them into PipeLineState.
+pub struct PipeLineGetter {
+    fifos: Vec<u32>,
+    rules: Vec<u32>,
+    b2r_cache: Arc<Mutex<HashMap<u32, VecDeque<B2RMessage>>>>,
+}
+
+impl PipeLineGetter {
+    pub fn new(server: &B2RServer) -> Self {
+        PipeLineGetter {
+            fifos: Vec::new(),
+            rules: Vec::new(),
+            b2r_cache: server.b2r_cache.clone(),
+        }
+    }
+
+    /// add a fifo probe
+    /// The fifo probe won't get data from rust, sent 2 bytes every cycle,
+    /// the fist byte is notFull second byte is notEmpty
+    pub fn add_fifo_probe(&mut self, id: u32) {
+        self.fifos.push(id);
+    }
+
+    /// add a rule probe
+    /// The rule probe won't get data from rust,
+    /// sent 1 bytes message when the rule fired
+    pub fn add_rule_probe(&mut self, id: u32) {
+        self.rules.push(id);
+    }
+
+    /// Read the earliest cycle messages sent by the probes labeled as "fifo" and "fired", and organize them into a PipeLineState.
+    pub fn get_pipeline_state(&mut self) -> PipeLineState {
+        let mut state: PipeLineState = PipeLineState {
+            cycle: u32::MAX,
+            full_fifos: Vec::new(),
+            empty_fifos: Vec::new(),
+            fire_rules: Vec::new(),
+        };
+
+        let mut b2r_cache: std::sync::MutexGuard<HashMap<u32, VecDeque<B2RMessage>>> =
+            self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+
+        for fifo_id in &self.fifos {
+            if let Some(messages) = b2r_cache.get(fifo_id) {
+                if let Some(first_message) = messages.front() {
+                    if first_message.cycles < state.cycle {
+                        state.cycle = first_message.cycles;
+                    }
+                }
+            }
+        }
+
+        for fifo_id in &self.fifos {
+            if let Some(messages) = b2r_cache.get_mut(fifo_id) {
+                if let Some(first_message) = messages.front() {
+                    if first_message.cycles == state.cycle {
+                        // the fifo message len must be 2
+                        assert_eq!(first_message.message.len(), 2);
+                        let b2r_message = messages.pop_front().expect("front error");
+                        if b2r_message.message[0] == 0 {
```

Comment:
No magic number here. You should define a struct to represent this kind of message, which has two fields, one for empty and one for full.

If this is a really big project, or message type boost in the future, it will be hard to remember the meaning of each byte index.